sklearn/calibration.py

```diff
@@ -25,6 +27,241 @@
+                                      self.pos_label,
+                                      self.min_tnr,
+                                      self.min_tpr).fit(
+                        X[test], y[test]
+                    ).threshold_
+                )
+            self.threshold_ = sum(thresholds) / len(thresholds)
+            self.base_estimator.fit(X, y)
+        return self
+
+    def predict(self, X):
+        """Predict using the calibrated decision threshold
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The samples.
+
+        Returns
+        -------
+        C : array, shape (n_samples,)
+            The predicted class.
+        """
+        X = check_array(X)
+        check_is_fitted(self, ["label_encoder_", "threshold_"])
+
+        return self.label_encoder_.inverse_transform(
+            (self.base_estimator.predict_proba(X)[:, self.pos_label] >
+             self.threshold_).astype(int)
+        )
+
+
+class _CutoffClassifier(object):
+    """Optimal cutoff point selection.
+
+    It assumes that base_estimator has already been fit, and uses the input set
+    of the fit function to select an optimal cutoff point. Note that this
+    class should not be used as an estimator directly. Use the
+    OptimalCutoffClassifier with cv="prefit" instead.
+
+    Parameters
+    ----------
+    base_estimator : instance BaseEstimator
+        The classifier whose decision threshold will be adapted according to
+        the acquired optimal cutoff point
+
+    method : 'roc' or 'max_tpr' or 'max_tnr'
+        The method to use for choosing the cutoff point.
+
+    pos_label : object
+        Label considered as positive during the roc_curve construction.
+
+    min_tnr : float in [0, 1]
+        minimum required value for true negative rate (specificity) in case
+        method 'max_tpr' is used
+
+    min_tpr : float in [0, 1]
+        minimum required value for true positive rate (sensitivity) in case
+        method 'max_tnr' is used
+
+    Attributes
+    ----------
+    threshold_ : float
+        Acquired optimal decision threshold for the positive class
+    """
+    def __init__(self, base_estimator, method, pos_label, min_tnr, min_tpr):
+        self.base_estimator = base_estimator
+        self.method = method
+        self.pos_label = pos_label
+        self.min_tnr = min_tnr
+        self.min_tpr = min_tpr
+
+    def fit(self, X, y):
+        """Select a decision threshold for the fitted model's positive class
+        using one of the available methods
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training data.
+
+        y : array-like, shape (n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : object
+            Instance of self.
+        """
+        y_score = self.base_estimator.predict_proba(X)[:, self.pos_label]
```

Comment:
can we not also use decision_function?

Reply:
yes, we can; I will add that. What should happen if a classifier has both decision function and predict proba?

I will make sure to also remove the base estimator check you mentioned bellow and move all validation in the beginning  of fit

Reply:
Follow the logic of [`_ThresholdScorer`](https://github.com/scikit-learn/scikit-learn/blob/e78263ff4a61a915fbcd65a62987369277296db4/sklearn/metrics/scorer.py#L153)